eZ/Bundle/EzPublishCoreBundle/Routing/DefaultRouter.php

```diff
@@ -95,7 +96,13 @@ public function generate($name, $parameters = array(), $referenceType = self::AB
             unset($parameters['siteaccess']);
         }
 
-        $url = parent::generate($name, $parameters, $referenceType);
+        try {
+            $url = parent::generate($name, $parameters, $referenceType);
+        } catch (RouteNotFoundException $e) {
+            // Switch back to original context, for next links generation.
+            $this->setContext($originalContext);
+            throw $e;
+        }
```

Comment:
From the OOP perspective I'd utilize `try-finally` in this case, wrapping the rest of the code like that: https://gist.github.com/alongosz/a6ca06dc34c7d8135503a3ee43751213
It even does not exceed ~100 character in line (more or less), so shouldn't really be too much of a "spaghetti" ;)